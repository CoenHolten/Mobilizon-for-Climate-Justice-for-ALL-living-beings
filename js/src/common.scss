a {
  text-decoration: underline;
  text-decoration-color: #fea72b;
  text-decoration-thickness: 2px;

  &.navbar-item,
  &.dropdown-item,
  &.card,
  &.button,
  &[href="#comments"],
  &.router-link-active,
  &.comment-link,
  &.pagination-link {
    text-decoration: none;
  }
}

input.input {
  border-color: $input-border-color !important;
}

.section {
  padding: 1rem 2rem 4rem;
}

figure img.is-rounded {
  border: 1px solid #cdcaea;
}
