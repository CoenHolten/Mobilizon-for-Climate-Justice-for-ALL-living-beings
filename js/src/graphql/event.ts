import gql from 'graphql-tag';

const participantQuery = `
  role,
  actor {
    preferredUsername,
    avatar {
      url
    },
    name,
    id
  }
`;

export const FETCH_EVENT = gql`
  query($uuid:UUID!) {
    event(uuid: $uuid) {
      id,
      uuid,
      url,
      local,
      title,
      slug,
      description,
      beginsOn,
      endsOn,
      status,
      visibility,
      picture {
        id
        url
      },
      publishAt,
      category,
      # online_address,
      # phone_address,
      physicalAddress {
        description,
        floor,
        street,
        locality,
        postalCode,
        region,
        country,
        geom
      }
      organizerActor {
        avatar {
          url
        },
        preferredUsername,
        domain,
        name,
      },
      # attributedTo {
      #     avatar {
      #      url,
      #     }
      #     preferredUsername,
      #     name,
      # },
      participants {
        ${participantQuery}
      },
      tags {
        slug,
        title
      },
      relatedEvents {
        uuid,
        title,
        beginsOn,
        physicalAddress {
          description
        },
        organizerActor {
          avatar {
            url,
          },
          preferredUsername,
          domain,
          name,
        }
      }
    }
  }
`;

export const FETCH_EVENTS = gql`
  query {
    events {
      id,
      uuid,
      url,
      local,
      title,
      description,
      beginsOn,
      endsOn,
      status,
      visibility,
      picture {
        id
        url
      },
      publishAt,
      # online_address,
      # phone_address,
      physicalAddress {
        description,
        locality
      }
      organizerActor {
        avatar {
          url
        },
        preferredUsername,
        name,
      },
      attributedTo {
        avatar {
          url
        },
        preferredUsername,
        name,
      },
      category,
      participants {
        ${participantQuery}
      },
      tags {
        slug,
        title
      },
    }
  }
`;

export const CREATE_EVENT = gql`
  mutation CreateEvent(
  $title: String!,
  $description: String!,
  $organizerActorId: ID!,
  $category: String!,
  $beginsOn: DateTime!,
  $picture_file: Upload,
  $picture_name: String,
  ) {
    createEvent(
      title: $title,
      description: $description,
      beginsOn: $beginsOn,
      organizerActorId: $organizerActorId,
      category: $category,
      picture: {
          picture: {
              file: $picture_file,
              name: $picture_name,
          }
      }
    ) {
      id,
      uuid,
      title,
      picture {
        url
      }
    }
  }
`;

export const EDIT_EVENT = gql`
  mutation EditEvent(
  $title: String!,
  $description: String!,
  $organizerActorId: Int!,
  $category: String!
  ) {
    EditEvent(title: $title, description: $description, organizerActorId: $organizerActorId, category: $category) {
      uuid
    }
  }
`;

export const JOIN_EVENT = gql`
  mutation JoinEvent($eventId: Int!, $actorId: Int!) {
    joinEvent(
      eventId: $eventId,
      actorId: $actorId
    ) {
        ${participantQuery}
    }
  }
`;

export const LEAVE_EVENT = gql`
  mutation LeaveEvent($eventId: Int!, $actorId: Int!) {
    leaveEvent(
      eventId: $eventId,
      actorId: $actorId
    ) {
      actor {
        id
      }
    }
  }
`;

export const DELETE_EVENT = gql`
  mutation DeleteEvent($id: Int!, $actorId: Int!) {
    deleteEvent(
      id: $id,
      actorId: $actorId
    )
  }
`;
