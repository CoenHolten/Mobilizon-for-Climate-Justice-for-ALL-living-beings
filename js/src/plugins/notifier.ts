import Vue from 'vue';

declare module 'vue/types/vue' {
  interface Vue {
    $notifier: {
      success: (message: string) => void;
    };
  }
}

export class Notifier {
  private readonly vue: typeof Vue;

  constructor(vue: typeof Vue) {
    this.vue = vue;
  }

  success(message: string) {
    this.vue.prototype.$notification.open({
      message,
      duration: 5000,
      position: 'is-bottom-right',
      type: 'is-success',
      hasIcon: true,
    });
  }
}

// tslint:disable
export function NotifierPlugin(vue: typeof Vue, options?: any): void {
  vue.prototype.$notifier = new Notifier(vue);
}
