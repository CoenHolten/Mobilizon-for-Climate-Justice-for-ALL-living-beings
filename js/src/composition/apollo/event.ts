import { DELETE_EVENT, FETCH_EVENT, FETCH_EVENT_BASIC } from "@/graphql/event";
import { IEvent } from "@/types/event.model";
import { useMutation, useQuery } from "@vue/apollo-composable";
import { Ref, computed, unref } from "vue";

export function useFetchEvent(uuidValue?: string | Ref<string>) {
  const uuid = unref(uuidValue);
  const {
    result: fetchEventResult,
    loading,
    error,
    onError,
    onResult,
    refetch,
  } = useQuery<{ event: IEvent }>(
    FETCH_EVENT,
    () => ({
      uuid,
    }),
    () => ({
      enabled: uuid !== undefined,
    })
  );

  const event = computed(() => fetchEventResult.value?.event);

  return { event, loading, error, onError, onResult, refetch };
}

export function useFetchEventBasic(uuidValue?: string | Ref<string>) {
  const uuid = unref(uuidValue);
  const {
    result: fetchEventResult,
    loading,
    error,
    onResult,
    onError,
  } = useQuery<{ event: IEvent }>(FETCH_EVENT_BASIC, {
    uuid,
  });

  const event = computed(() => fetchEventResult.value?.event);

  return { event, loading, error, onResult, onError };
}

export function useDeleteEvent() {
  return useMutation<{ id: string }, { eventId: string }>(DELETE_EVENT, () => ({
    update(cache, { data }) {
      cache.evict({ id: `Event:${data?.id}` });
      cache.gc();
    },
  }));
}
