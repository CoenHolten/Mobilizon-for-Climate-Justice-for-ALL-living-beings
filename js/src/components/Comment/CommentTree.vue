<template>
  <div>
    <form
      class=""
      v-if="isAbleToComment"
      @submit.prevent="createCommentForEvent(newComment)"
      @keyup.ctrl.enter="createCommentForEvent(newComment)"
    >
      <o-notification
        v-if="isEventOrganiser && !areCommentsClosed"
        :closable="false"
        class="my-2"
        >{{ t("Comments are closed for everybody else.") }}</o-notification
      >
      <article class="flex flex-wrap items-start gap-2">
        <figure class="" v-if="newComment.actor">
          <identity-picker-wrapper :inline="false" v-model="newComment.actor" />
        </figure>
        <div class="flex-1">
          <div class="flex flex-col gap-2">
            <div class="editor-wrapper">
              <Editor
                ref="commenteditor"
                v-if="currentActor"
                :currentActor="currentActor"
                mode="comment"
                v-model="newComment.text"
                :aria-label="t('Comment body')"
              />
              <p class="" v-if="emptyCommentError">
                {{ t("Comment text can't be empty") }}
              </p>
            </div>
            <div class="" v-if="isEventOrganiser">
              <o-switch
                aria-labelledby="notify-participants-toggle"
                v-model="newComment.isAnnouncement"
                >{{ t("Notify participants") }}</o-switch
              >
            </div>
          </div>
        </div>
        <div class="">
          <o-button native-type="submit" variant="primary" icon-left="send">{{
            t("Send")
          }}</o-button>
        </div>
      </article>
    </form>
    <o-notification v-else-if="isConnected" :closable="false">{{
      t("The organiser has chosen to close comments.")
    }}</o-notification>
    <p v-if="commentsLoading" class="text-center">
      {{ t("Loading commentsâ€¦") }}
    </p>
    <transition-group tag="div" name="comment-empty-list" v-else>
      <transition-group
        key="list"
        name="comment-list"
        v-if="filteredOrderedComments.length && currentActor"
        class="comment-list"
        tag="ul"
      >
        <comment
          class="root-comment"
          :comment="comment"
          :event="event"
          :currentActor="currentActor"
          v-for="comment in filteredOrderedComments"
          :key="comment.id"
          @create-comment="createCommentForEvent"
          @delete-comment="
            deleteComment({
              commentId: comment.id as string,
              originCommentId: comment.originComment?.id,
            })
          "
        />
      </transition-group>
      <empty-content v-else icon="comment" key="no-comments" :inline="true">
        <span>{{ t("No comments yet") }}</span>
      </empty-content>
    </transition-group>
  </div>
</template>

<script lang="ts" setup>
import Comment from "@/components/Comment/Comment.vue";
import IdentityPickerWrapper from "@/views/Account/IdentityPickerWrapper.vue";
import { CommentModeration } from "@/types/enums";
import { CommentModel, IComment } from "../../types/comment.model";
import {
  CREATE_COMMENT_FROM_EVENT,
  DELETE_COMMENT,
  COMMENTS_THREADS_WITH_REPLIES,
} from "../../graphql/comment";
import { IEvent } from "../../types/event.model";
import { ApolloCache, FetchResult, InMemoryCache } from "@apollo/client/core";
import EmptyContent from "@/components/Utils/EmptyContent.vue";
import { useCurrentActorClient } from "@/composition/apollo/actor";
import { useMutation, useQuery } from "@vue/apollo-composable";
import { computed, defineAsyncComponent, inject, ref, watch } from "vue";
import { IPerson } from "@/types/actor";
import { AbsintheGraphQLError } from "@/types/errors.model";
import { useI18n } from "vue-i18n";
import { Notifier } from "@/plugins/notifier";

const { currentActor } = useCurrentActorClient();

const { result: commentsResult, loading: commentsLoading } = useQuery<{
  event: Pick<IEvent, "id" | "uuid" | "comments">;
}>(
  COMMENTS_THREADS_WITH_REPLIES,
  () => ({ eventUUID: props.event?.uuid }),
  () => ({ enabled: props.event?.uuid !== undefined })
);

const comments = computed(() => commentsResult.value?.event.comments ?? []);

const props = defineProps<{
  event: IEvent;
  newComment?: IComment;
}>();

const Editor = defineAsyncComponent(() => import("@/components/Editor.vue"));

const newComment = ref<IComment>(props.newComment ?? new CommentModel());

const emptyCommentError = ref(false);

const { t } = useI18n({ useScope: "global" });

watch(currentActor, () => {
  newComment.value.actor = currentActor.value as IPerson;
});

watch(newComment, (newCommentUpdated: IComment) => {
  if (emptyCommentError.value) {
    emptyCommentError.value = ["", "<p></p>"].includes(newCommentUpdated.text);
  }
});

const {
  mutate: createCommentForEventMutation,
  onDone: createCommentForEventMutationDone,
  onError: createCommentForEventMutationError,
} = useMutation<
  { createComment: IComment },
  {
    eventId: string;
    text: string;
    inReplyToCommentId?: string;
    isAnnouncement?: boolean;
    originCommentId?: string | undefined;
  }
>(CREATE_COMMENT_FROM_EVENT, () => ({
  update: (
    store: ApolloCache<InMemoryCache>,
    { data }: FetchResult,
    { variables }
  ) => {
    if (data == null) return;
    // comments are attached to the event, so we can pass it to replies later
    const newCommentLocal = { ...data.createComment, event: props.event };

    // we load all existing threads
    const commentThreadsData = store.readQuery<{ event: IEvent }>({
      query: COMMENTS_THREADS_WITH_REPLIES,
      variables: {
        eventUUID: props.event?.uuid,
      },
    });
    if (!commentThreadsData) return;
    const { event } = commentThreadsData;
    const oldComments = [...event.comments];

    // if it's no a root comment, we first need to find
    // existing replies and add the new reply to it
    if (variables?.originCommentId !== undefined) {
      const parentCommentIndex = oldComments.findIndex(
        (oldComment) => oldComment.id === variables.originCommentId
      );
      const parentComment = oldComments[parentCommentIndex];

      // replace the root comment with has the updated list of replies in the thread list
      oldComments.splice(parentCommentIndex, 1, {
        ...parentComment,
        replies: [...parentComment.replies, newCommentLocal],
      });
    } else {
      // otherwise it's simply a new thread and we add it to the list
      oldComments.push(newCommentLocal);
    }

    // finally we save the thread list
    store.writeQuery({
      query: COMMENTS_THREADS_WITH_REPLIES,
      data: {
        event: {
          ...event,
          comments: oldComments,
        },
      },
      variables: {
        eventUUID: props.event?.uuid,
      },
    });
  },
}));

createCommentForEventMutationDone(() => {
  // and reset the new comment field
  newComment.value = new CommentModel();
});

const notifier = inject<Notifier>("notifier");

createCommentForEventMutationError((errors) => {
  console.error(errors);
  if (errors.graphQLErrors && errors.graphQLErrors.length > 0) {
    const error = errors.graphQLErrors[0] as AbsintheGraphQLError;
    if (error.field !== "text" && error.message[0] !== "can't be blank") {
      notifier?.error(error.message);
    }
  }
});

const createCommentForEvent = (comment: IComment) => {
  emptyCommentError.value = ["", "<p></p>"].includes(comment.text);

  if (emptyCommentError.value) return;
  if (!comment.actor) return;
  if (!props.event?.id) return;

  createCommentForEventMutation({
    eventId: props.event?.id,
    text: comment.text,
    inReplyToCommentId: comment.inReplyToComment?.id,
    isAnnouncement: comment.isAnnouncement,
    originCommentId: comment.originComment?.id,
  });
};

const { mutate: deleteComment, onError: deleteCommentMutationError } =
  useMutation<
    { deleteComment: { id: string } },
    { commentId: string; originCommentId?: string }
  >(DELETE_COMMENT, () => ({
    update: (
      store: ApolloCache<InMemoryCache>,
      { data }: FetchResult,
      { variables }
    ) => {
      if (data == null) return;
      const deletedCommentId = data.deleteComment.id;

      const commentsData = store.readQuery<{ event: IEvent }>({
        query: COMMENTS_THREADS_WITH_REPLIES,
        variables: {
          eventUUID: props.event?.uuid,
        },
      });
      if (!commentsData) return;
      const { event } = commentsData;
      let updatedComments: IComment[] = [...event.comments];

      if (variables?.originCommentId) {
        // we have deleted a reply to a thread
        const parentCommentIndex = updatedComments.findIndex(
          (oldComment) => oldComment.id === variables.originCommentId
        );
        const parentComment = updatedComments[parentCommentIndex];
        const updatedReplies = parentComment.replies.map((reply) => {
          if (reply.id === deletedCommentId) {
            return {
              ...reply,
              deletedAt: new Date().toString(),
            };
          }
          return reply;
        });
        updatedComments.splice(parentCommentIndex, 1, {
          ...parentComment,
          replies: updatedReplies,
          totalReplies: parentComment.totalReplies - 1,
        });
        console.log("updatedComments", updatedComments);
      } else {
        // we have deleted a thread itself
        updatedComments = updatedComments.map((reply) => {
          if (reply.id === deletedCommentId) {
            return {
              ...reply,
              deletedAt: new Date().toString(),
            };
          }
          return reply;
        });
      }
      store.writeQuery({
        query: COMMENTS_THREADS_WITH_REPLIES,
        variables: {
          eventUUID: props.event?.uuid,
        },
        data: {
          event: {
            ...event,
            comments: updatedComments,
          },
        },
      });
    },
  }));

deleteCommentMutationError((error) => {
  console.error(error);
  if (error.graphQLErrors && error.graphQLErrors.length > 0) {
    notifier?.error(error.graphQLErrors[0].message);
  }
});

const orderedComments = computed((): IComment[] => {
  return comments.value
    .filter((comment: IComment) => comment.inReplyToComment == null)
    .sort((a: IComment, b: IComment) => {
      if (a.isAnnouncement !== b.isAnnouncement) {
        return (
          (b.isAnnouncement === true ? 1 : 0) -
          (a.isAnnouncement === true ? 1 : 0)
        );
      }
      if (a.publishedAt && b.publishedAt) {
        return (
          new Date(b.publishedAt).getTime() - new Date(a.publishedAt).getTime()
        );
      } else if (a.updatedAt && b.updatedAt) {
        return (
          new Date(b.updatedAt).getTime() - new Date(a.updatedAt).getTime()
        );
      }
      return 0;
    });
});

const filteredOrderedComments = computed((): IComment[] => {
  return orderedComments.value.filter(
    (comment) => !comment.deletedAt || comment.totalReplies > 0
  );
});

const isEventOrganiser = computed((): boolean => {
  const organizerId =
    props.event?.organizerActor?.id || props.event?.attributedTo?.id;
  return organizerId !== undefined && currentActor.value?.id === organizerId;
});

const areCommentsClosed = computed((): boolean => {
  return (
    currentActor.value?.id !== undefined &&
    props.event?.options.commentModeration !== CommentModeration.CLOSED
  );
});

const isAbleToComment = computed((): boolean => {
  if (isConnected.value) {
    return areCommentsClosed.value || isEventOrganiser.value;
  }
  return false;
});

const isConnected = computed((): boolean => {
  return currentActor.value?.id != undefined;
});
</script>

<style lang="scss" scoped>
// @use "@/styles/_mixins" as *;
// // @import "node_modules/bulma/sass/utilities/mixins.sass";
// form.new-comment {
//   padding-bottom: 1rem;

//   .media {
//     flex-wrap: wrap;
//     justify-content: center;
//     // .media-left {
//     //   @include >mobile {
//     //     @include margin-right(0.5rem);
//     //     @include margin-left(0.5rem);
//     //   }
//     // }

//     .media-content {
//       display: flex;
//       align-items: center;
//       align-content: center;
//       width: min-content;

//       .field {
//         flex: 1;
//         // @include padding-right(10px);
//         margin-bottom: 0;

//         &.notify-participants {
//           margin-top: 0.5rem;
//         }
//       }
//     }
//   }
// }

// .no-comments {
//   display: flex;
//   flex-direction: column;

//   span {
//     text-align: center;
//     margin-bottom: 10px;
//   }

//   img {
//     max-width: 250px;
//     align-self: center;
//   }
// }

// ul.comment-list li {
//   margin-bottom: 16px;
// }

.comment-list-enter-active,
.comment-list-leave-active,
.comment-list-move {
  transition: 500ms cubic-bezier(0.59, 0.12, 0.34, 0.95);
  transition-property: opacity, transform;
}

.comment-list-enter {
  opacity: 0;
  transform: translateX(50px) scaleY(0.5);
}

.comment-list-enter-to {
  opacity: 1;
  transform: translateX(0) scaleY(1);
}

.comment-list-leave-active,
.comment-empty-list-active {
  position: absolute;
}

.comment-list-leave-to,
.comment-empty-list-leave-to {
  opacity: 0;
  transform: scaleY(0);
  transform-origin: center top;
}

// .comment-empty-list-enter-active {
//     transition: opacity .5s;
// }

// .comment-empty-list-enter {
//     opacity: 0;
// }
</style>
