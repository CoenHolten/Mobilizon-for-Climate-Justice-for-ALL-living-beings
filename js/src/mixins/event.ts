import { mixins } from "vue-class-component";
import { Component, Vue } from "vue-property-decorator";
import { IEvent, IParticipant, ParticipantRole } from "../types/event.model";
import {
  DELETE_EVENT,
  EVENT_PERSON_PARTICIPATION,
  FETCH_EVENT,
  LEAVE_EVENT,
} from "../graphql/event";
import RouteName from "../router/name";
import { IPerson } from "../types/actor";

@Component
export default class EventMixin extends mixins(Vue) {
  protected async leaveEvent(
    event: IEvent,
    actorId: string,
    token: string | null = null,
    anonymousParticipationConfirmed: boolean | null = null
  ) {
    try {
      const { data: resultData } = await this.$apollo.mutate<{ leaveEvent: IParticipant }>({
        mutation: LEAVE_EVENT,
        variables: {
          eventId: event.id,
          actorId,
          token,
        },
        update: (store, { data }) => {
          if (data == null) return;
          let participation;

          if (!token) {
            const participationCachedData = store.readQuery<{ person: IPerson }>({
              query: EVENT_PERSON_PARTICIPATION,
              variables: { eventId: event.id, actorId },
            });
            if (participationCachedData == null) return;
            const { person } = participationCachedData;
            if (person === null) {
              console.error("Cannot update participation cache, because of null value.");
              return;
            }
            [participation] = person.participations;
            person.participations = [];
            store.writeQuery({
              query: EVENT_PERSON_PARTICIPATION,
              variables: { eventId: event.id, actorId },
              data: { person },
            });
          }

          const eventCachedData = store.readQuery<{ event: IEvent }>({
            query: FETCH_EVENT,
            variables: { uuid: event.uuid },
          });
          if (eventCachedData == null) return;
          const { event: eventCached } = eventCachedData;
          if (eventCached === null) {
            console.error("Cannot update event cache, because of null value.");
            return;
          }
          if (participation && participation.role === ParticipantRole.NOT_APPROVED) {
            eventCached.participantStats.notApproved -= 1;
          } else if (anonymousParticipationConfirmed === false) {
            eventCached.participantStats.notConfirmed -= 1;
          } else {
            eventCached.participantStats.going -= 1;
            eventCached.participantStats.participant -= 1;
          }
          store.writeQuery({
            query: FETCH_EVENT,
            variables: { uuid: event.uuid },
            data: { event: eventCached },
          });
        },
      });
      if (resultData) {
        this.participationCancelledMessage();
      }
    } catch (error) {
      console.error(error);
    }
  }

  private participationCancelledMessage() {
    this.$notifier.success(this.$t("You have cancelled your participation") as string);
  }

  protected async openDeleteEventModal(event: IEvent, currentActor: IPerson) {
    function escapeRegExp(string: string) {
      return string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'); // $& means the whole matched string
    }  	  
    const participantsLength = event.participantStats.participant;
    const prefix = participantsLength
      ? this.$tc("There are {participants} participants.", event.participantStats.participant, {
          participants: event.participantStats.participant,
        })
      : "";

    this.$buefy.dialog.prompt({
      type: "is-danger",
      title: this.$t("Delete event") as string,
      message: `${prefix}
        ${this.$t("Are you sure you want to delete this event? This action cannot be reverted.")}
        <br><br>
        ${this.$t('To confirm, type your event title "{eventTitle}"', {
          eventTitle: event.title,
        })}`,
      confirmText: this.$t("Delete {eventTitle}", { eventTitle: event.title }) as string,
      inputAttrs: {
        placeholder: event.title,
        pattern: escapeRegExp(event.title),
      },
      onConfirm: () => this.deleteEvent(event, currentActor),
    });
  }

  private async deleteEvent(event: IEvent, currentActor: IPerson) {
    const eventTitle = event.title;

    try {
      await this.$apollo.mutate<IParticipant>({
        mutation: DELETE_EVENT,
        variables: {
          eventId: event.id,
          actorId: currentActor.id,
        },
      });
      /**
       * When the event corresponding has been deleted (by the organizer).
       * A notification is already triggered.
       *
       * @type {string}
       */
      this.$emit("eventDeleted", event.id);

      this.$buefy.notification.open({
        message: this.$t("Event {eventTitle} deleted", { eventTitle }) as string,
        type: "is-success",
        position: "is-bottom-right",
        duration: 5000,
      });
      await this.$router.push({ name: RouteName.HOME });
    } catch (error) {
      console.error(error);
    }
  }

  urlToHostname(url: string): string | null {
    try {
      return new URL(url).hostname;
    } catch (e) {
      return null;
    }
  }
}
