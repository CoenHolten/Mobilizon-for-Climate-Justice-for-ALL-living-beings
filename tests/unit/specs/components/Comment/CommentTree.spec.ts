import { config, shallowMount, VueWrapper } from "@vue/test-utils";
import CommentTree from "@/components/Comment/CommentTree.vue";
import {
  createMockClient,
  MockApolloClient,
  RequestHandler,
} from "mock-apollo-client";
import {
  COMMENTS_THREADS_WITH_REPLIES,
  CREATE_COMMENT_FROM_EVENT,
} from "@/graphql/comment";
import { CommentModeration } from "@/types/enums";
import { IEvent } from "@/types/event.model";
import {
  eventCommentThreadsMock,
  eventNoCommentThreadsMock,
  newCommentForEventMock,
  newCommentForEventResponse,
} from "../../mocks/event";
import flushPromises from "flush-promises";
import { defaultResolvers } from "../../common";
import { afterEach, describe, vi, it, expect, beforeEach } from "vitest";
import { DefaultApolloClient } from "@vue/apollo-composable";
import Oruga from "@oruga-ui/oruga-next";
import { notifierPlugin } from "@/plugins/notifier";
import { InMemoryCache } from "@apollo/client/cache";
import { createRouter, createWebHistory, Router } from "vue-router";
import { routes } from "@/router";
import { dialogPlugin } from "@/plugins/dialog";

config.global.plugins.push(Oruga);
config.global.plugins.push(notifierPlugin);
config.global.plugins.push(dialogPlugin);

let router: Router;

const eventData = {
  id: "1",
  uuid: "e37910ea-fd5a-4756-7634-00971f3f4107",
  options: {
    commentModeration: CommentModeration.ALLOW_ALL,
  },
};
describe("CommentTree", () => {
  let wrapper: VueWrapper;
  let mockClient: MockApolloClient | null;
  let requestHandlers: Record<string, RequestHandler>;

  const generateWrapper = (handlers = {}, extraProps = {}) => {
    const cache = new InMemoryCache({ addTypename: true });

    mockClient = createMockClient({
      cache,
      resolvers: defaultResolvers,
    });

    requestHandlers = {
      eventCommentThreadsQueryHandler: vi
        .fn()
        .mockResolvedValue(eventCommentThreadsMock),
      createCommentForEventMutationHandler: vi
        .fn()
        .mockResolvedValue(newCommentForEventResponse),
      ...handlers,
    };

    mockClient.setRequestHandler(
      COMMENTS_THREADS_WITH_REPLIES,
      requestHandlers.eventCommentThreadsQueryHandler
    );
    mockClient.setRequestHandler(
      CREATE_COMMENT_FROM_EVENT,
      requestHandlers.createCommentForEventMutationHandler
    );
    wrapper = shallowMount(CommentTree, {
      props: {
        event: { ...eventData },
        ...extraProps,
      },
      global: {
        provide: {
          [DefaultApolloClient]: mockClient,
        },
        plugins: [router],
      },
    });
  };

  beforeEach(async () => {
    router = createRouter({
      history: createWebHistory(),
      routes: routes,
    });

    // await router.isReady();
  });

  afterEach(() => {
    mockClient = null;
    requestHandlers = {};
    wrapper.unmount();
  });

  it("renders a loading comment tree", async () => {
    generateWrapper();

    expect(wrapper.find("p.text-center").text()).toBe("Loading commentsâ€¦");

    expect(wrapper.html()).toMatchSnapshot();
  });

  it("renders a comment tree with comments", async () => {
    generateWrapper();

    expect(wrapper.exists()).toBe(true);
    expect(
      requestHandlers.eventCommentThreadsQueryHandler
    ).toHaveBeenCalledWith({ eventUUID: eventData.uuid });
    await flushPromises();
    expect(wrapper.find("p.text-center").exists()).toBe(false);

    expect(wrapper.findAllComponents("event-comment-stub").length).toBe(2);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("renders the form if we can comment", async () => {
    generateWrapper(
      {},
      {
        newComment: {
          text: newCommentForEventMock.text,
          isAnnouncement: false,
          insertedAt: "2021-12-03T13:02:00Z",
          updatedAt: "2021-12-03T13:02:00Z",
          publishedAt: "2021-12-03T13:02:00Z",
        },
      }
    );

    await flushPromises();

    expect(wrapper.find("form").isVisible()).toBe(true);
    expect(wrapper.findAllComponents("event-comment-stub").length).toBe(2);
    wrapper.getComponent({ ref: "commenteditor" });

    wrapper.find("form").trigger("submit");

    await flushPromises();
    expect(
      requestHandlers.createCommentForEventMutationHandler
    ).toHaveBeenCalledWith({
      ...newCommentForEventMock,
    });

    if (mockClient) {
      const cachedData = mockClient.cache.readQuery<{ event: IEvent }>({
        query: COMMENTS_THREADS_WITH_REPLIES,
        variables: {
          eventUUID: eventData.uuid,
        },
      });
      if (cachedData) {
        const { event } = cachedData;

        expect(event.comments).toHaveLength(3);
      }
    }
  });

  it("renders an empty comment tree", async () => {
    generateWrapper({
      eventCommentThreadsQueryHandler: vi
        .fn()
        .mockResolvedValue(eventNoCommentThreadsMock),
    });

    await flushPromises();
    expect(
      requestHandlers.eventCommentThreadsQueryHandler
    ).toHaveBeenCalledWith({ eventUUID: eventData.uuid });

    expect(wrapper.findComponent({ name: "EmptyContent" }).exists());
    expect(wrapper.html()).toMatchSnapshot();
  });
});
