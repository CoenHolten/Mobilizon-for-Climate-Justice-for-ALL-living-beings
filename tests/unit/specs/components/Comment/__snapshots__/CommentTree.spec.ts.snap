// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CommentTree > renders a comment tree with comments 1`] = `
"<div data-v-5d0380ab=\\"\\">
  <form data-v-5d0380ab=\\"\\" class=\\"mt-2\\">
    <!--v-if-->
    <article data-v-5d0380ab=\\"\\" class=\\"flex flex-wrap items-start gap-2\\">
      <figure data-v-5d0380ab=\\"\\" class=\\"\\">
        <identity-picker-wrapper-stub data-v-5d0380ab=\\"\\" modelvalue=\\"[object Object]\\" inline=\\"false\\" masked=\\"false\\"></identity-picker-wrapper-stub>
      </figure>
      <div data-v-5d0380ab=\\"\\" class=\\"flex-1\\">
        <div data-v-5d0380ab=\\"\\" class=\\"flex flex-col gap-2\\">
          <div data-v-5d0380ab=\\"\\" class=\\"editor-wrapper\\">
            <editor-stub data-v-5d0380ab=\\"\\" currentactor=\\"[object Object]\\" mode=\\"comment\\" modelvalue=\\"\\" aria-label=\\"Comment body\\" placeholder=\\"Write a new comment\\"></editor-stub>
            <!--v-if-->
          </div>
          <!--v-if-->
        </div>
      </div>
      <div data-v-5d0380ab=\\"\\" class=\\"\\">
        <o-button-stub data-v-5d0380ab=\\"\\" variant=\\"primary\\" iconleft=\\"send\\" rounded=\\"false\\" outlined=\\"false\\" loading=\\"false\\" expanded=\\"false\\" inverted=\\"false\\" nativetype=\\"submit\\" tag=\\"button\\" disabled=\\"false\\"></o-button-stub>
      </div>
    </article>
  </form>
  <transition-group-stub data-v-5d0380ab=\\"\\" tag=\\"div\\" name=\\"comment-empty-list\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\" class=\\"mt-2\\">
    <transition-group-stub data-v-5d0380ab=\\"\\" name=\\"comment-list\\" tag=\\"ul\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\" class=\\"comment-list\\">
      <event-comment-stub data-v-5d0380ab=\\"\\" comment=\\"[object Object]\\" event=\\"[object Object]\\" currentactor=\\"[object Object]\\" rootcomment=\\"true\\" readonly=\\"false\\" class=\\"root-comment my-2\\"></event-comment-stub>
      <event-comment-stub data-v-5d0380ab=\\"\\" comment=\\"[object Object]\\" event=\\"[object Object]\\" currentactor=\\"[object Object]\\" rootcomment=\\"true\\" readonly=\\"false\\" class=\\"root-comment my-2\\"></event-comment-stub>
    </transition-group-stub>
  </transition-group-stub>
</div>"
`;

exports[`CommentTree > renders a loading comment tree 1`] = `
"<div data-v-5d0380ab=\\"\\">
  <!--v-if-->
  <p data-v-5d0380ab=\\"\\" class=\\"text-center\\">Loading commentsâ€¦</p>
</div>"
`;

exports[`CommentTree > renders an empty comment tree 1`] = `
"<div data-v-5d0380ab=\\"\\">
  <form data-v-5d0380ab=\\"\\" class=\\"mt-2\\">
    <!--v-if-->
    <article data-v-5d0380ab=\\"\\" class=\\"flex flex-wrap items-start gap-2\\">
      <figure data-v-5d0380ab=\\"\\" class=\\"\\">
        <identity-picker-wrapper-stub data-v-5d0380ab=\\"\\" modelvalue=\\"[object Object]\\" inline=\\"false\\" masked=\\"false\\"></identity-picker-wrapper-stub>
      </figure>
      <div data-v-5d0380ab=\\"\\" class=\\"flex-1\\">
        <div data-v-5d0380ab=\\"\\" class=\\"flex flex-col gap-2\\">
          <div data-v-5d0380ab=\\"\\" class=\\"editor-wrapper\\">
            <editor-stub data-v-5d0380ab=\\"\\" currentactor=\\"[object Object]\\" mode=\\"comment\\" modelvalue=\\"\\" aria-label=\\"Comment body\\" placeholder=\\"Write a new comment\\"></editor-stub>
            <!--v-if-->
          </div>
          <!--v-if-->
        </div>
      </div>
      <div data-v-5d0380ab=\\"\\" class=\\"\\">
        <o-button-stub data-v-5d0380ab=\\"\\" variant=\\"primary\\" iconleft=\\"send\\" rounded=\\"false\\" outlined=\\"false\\" loading=\\"false\\" expanded=\\"false\\" inverted=\\"false\\" nativetype=\\"submit\\" tag=\\"button\\" disabled=\\"false\\"></o-button-stub>
      </div>
    </article>
  </form>
  <transition-group-stub data-v-5d0380ab=\\"\\" tag=\\"div\\" name=\\"comment-empty-list\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\" class=\\"mt-2\\">
    <empty-content-stub data-v-5d0380ab=\\"\\" icon=\\"comment\\" descriptionclasses=\\"\\" inline=\\"true\\" center=\\"false\\"></empty-content-stub>
  </transition-group-stub>
</div>"
`;
